config/routes.rb

```diff
@@ -16,18 +16,31 @@
 
   resources :messages, only: [:show, :update]
   resources :documents
-  resources :users, only: [:new, :create, :update]
+  constraints lambda { |_| Feature.authenticate_with_password? } do
+    resources :users, only: [:new, :create, :update]
+  end
+  constraints lambda { |_| Feature.authenticate_with_google? || Feature.authenticate_with_http_header? } do
+    resources :users, only: [:update]
+  end
 
   get "up" => "rails/health#show", as: :rails_health_check
 
   # routes to still be cleaned up:
 
   resources :chats, only: [:index, :show, :create]
 
-  get "/login", to: "sessions#new"
-  post "/login", to: "sessions#create"
-  get "/register", to: "users#new"
-  get "/logout", to: "sessions#destroy"
+  constraints lambda { |_| Feature.authenticate_with_password? } do
+    get "/login", to: "sessions/passwords#new"
+    post "/login", to: "sessions/passwords#create"
+    get "/register", to: "users#new"
+    get "/logout", to: "sessions/passwords#destroy"
+  end
+
+  constraints lambda { |_| Feature.authenticate_with_google? } do
+    get "/login", to: "sessions/google#new"
+    get "/auth/google/callback", to: "sessions/google#create"
+    get "/logout", to: "sessions/google#destroy"
+  end
```

Comment:
I don’t love the complexity of the routes & controllers now. I’m staring at this and trying to think of how it can be simplified so it’s not so much to wrap your mind around. I don’t have a great suggestion yet.

Do we need to conditionalize routes? What if we just set up all the routes and then within them we have before_filters that forward if you shouldn’t be allowed to use it. I think that may be a simpler convention.

Also, do we really need to namespace controllers? Something smells off about that. For example, PasswordsController doesn’t feel right. Password isn’t the model we’re concerned with: we’re Creating / Reading / Updating / Deleting passwords. Like, we’re not Creating new Google’s (which is what a GoogleController would imply). I’ll think more about this …

Reply:
It was an idea to isolate the logic. But I agree, it is not that nice. I will change it.